### PR TITLE
feat(firebase): add Google Account auth

### PR DESCRIPTION
**Pull request by penx**
Reopening https://github.com/remix-run/examples/pull/4, https://github.com/remix-run/remix/pull/4057 against remix-run/examples

Firebase example - Integration with Google Sign-in Provider

Closes: "sign-in with Google" suggestion at https://github.com/remix-run/remix/pull/1811#issuecomment-1050238130

Adds a "Login with Google" link:

> Screenshot 2022-08-24 at 15 16 57

...that takes a user to the Google's auth screen:

> Screenshot 2022-08-24 at 15 17 22

 ...and then creates an account (if it doesn't yet exist) and signs them in:

> Screenshot 2022-08-24 at 15 19 26

### CODE REVIEW

**Comment by penx** [comment]
@machour @mcansh @MichaelDeBoey 

happy to abandon this if it's not wanted, but I've opened this PR 3 times now, and rebased several times on request, now github actions is telling me you're waiting for a response from me 😄

File: firebase/app/routes/auth/google.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import type { LoaderFunction } from "@remix-run/node";
+import { redirect } from "@remix-run/node";
+
+import { signInWithIdp } from "~/server/auth.server";
+import { commitSession, getSession } from "~/sessions";
+
+export const loader: LoaderFunction = async ({ request }) => {
```
Comment (MichaelDeBoey) [inline firebase/app/routes/auth/google.tsx:7]:
```suggestion
export const loader = async ({ request }: LoaderArgs) => {
```

File: firebase/app/routes/auth/google.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import type { LoaderFunction } from "@remix-run/node";
```
Comment (MichaelDeBoey) [inline firebase/app/routes/auth/google.tsx:1]:
```suggestion
import type { LoaderArgs } from "@remix-run/node";
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]
@penx Please rebase this branch onto latest `dev` & resolve conflicts and remarks

**Comment by penx** [comment]
@MichaelDeBoey rebased and confirmed it's all still working. Also addressed feedback and updated the docs a little.

File: firebase/README.md [hunk 3]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
 
 ## Running locally
 
-To run it, you need to either:
+To run this example, you either need to create a firebase proejct or use the emulators:
```
Comment (MichaelDeBoey) [inline firebase/README.md]:
```suggestion
To run this example, you either need to create a Firebase project or use [the emulators](https://firebase.google.com/docs/emulator-suite):
```

Comment (penx) [inline firebase/README.md:15]:
```suggestion
To run this example, you either need to create a firebase project or use the emulators:
```
```suggestion
To run this example, you either need to create a firebase proejct or use the emulators:
```

File: firebase/README.md [hunk 4]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
 
 ## Running locally
 
-To run it, you need to either:
+To run this example, you either need to create a Firebase project or use the emulators:
```
Comment (MichaelDeBoey) [inline firebase/README.md:15]:
```suggestion
To run this example, you either need to create a Firebase project or use [the emulators](https://firebase.google.com/docs/emulator-suite):
```

File: firebase/README.md [hunk 5]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
...
 
 ### 1. Run against a Firebase Project
```
Comment (MichaelDeBoey) [inline firebase/README.md:17]:
```suggestion
### 1. Run against a Firebase project
```

File: firebase/app/routes/login.tsx [hunk 6]
```diff
@@ -113,6 +118,17 @@ export default function Login() {
...
+        <a
+          href={`https://accounts.google.com/o/oauth2/v2/auth\
+?response_type=code\
+&client_id=${GOOGLE_CLIENT_ID}\
+&redirect_uri=http://localhost:3000/auth/google\
+&scope=openid%20email%20profile`}
+        >
+          Login with Google
+        </a>
+      </p>
```
Comment (penx) [inline firebase/app/routes/login.tsx:131]: Could perhaps hide this if GOOGLE_CLIENT_ID is not set

File: firebase/README.md [hunk 7]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
...
 
-1. [Create a Firebase Project](https://console.firebase.google.com)
-2. Enable Auth (with email) and Firestore
-3. Add a Web App
-4. Get the [admin-sdk](https://firebase.google.com/docs/admin/setup#initialize-sdk) and [Web API Key](https://firebase.google.com/docs/reference/rest/auth)
-5. Save them to SERVICE_ACCOUNT and API_KEY in the `.env`-file
+1. [Create a Firebase Project](https://console.firebase.google.com).
```
Comment (MichaelDeBoey) [inline firebase/README.md:19]:
```suggestion
1. [Create a Firebase project](https://console.firebase.google.com).
```

File: firebase/app/routes/auth/google.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  const response = await fetch("https://oauth2.googleapis.com/token", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      client_id: process.env.GOOGLE_CLIENT_ID,
+      client_secret: process.env.GOOGLE_CLIENT_SECRET,
+      code,
+      redirect_uri: "http://localhost:3000/auth/google",
```
Comment (penx) [inline firebase/app/routes/auth/google.tsx:20]: I don't remember why this was hard coded to localhost. Perhaps something is needed here. Will need to test a deployed version.

File: firebase/README.md [hunk 9]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
...
+2. Enable Auth (with email) and Firestore.
```
Comment (MichaelDeBoey) [inline firebase/README.md:20]:
```suggestion
2. Enable [Auth](https://firebase.google.com/docs/auth) (with email) and [Cloud Firestore](https://firebase.google.com/docs/firestore).
```

File: firebase/README.md [hunk 10]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
...
+3. Add a Web App with hosting.
+4. Copy `.env.example` to `.env`.
+5. Set the API_KEY value to the [Web API Key](https://firebase.google.com/docs/reference/rest/auth), which can be obtained on the project settings page in your Firebase admin console.
```
Comment (MichaelDeBoey) [inline firebase/README.md:23]:
```suggestion
5. Set the `API_KEY` value to the [Web API Key](https://firebase.google.com/docs/reference/rest/auth), which can be obtained on the project settings page in your Firebase admin console.
```

File: firebase/README.md [hunk 11]
```diff
@@ -12,15 +12,18 @@ See the screen recording at `./screen_recording.gif` or Open this example on [Co
...
+6. Set SERVICE_ACCOUNT to the contents of your service account's private key JSON file:
```
Comment (MichaelDeBoey) [inline firebase/README.md:24]:
```suggestion
6. Set `SERVICE_ACCOUNT` to the contents of your service account's private key JSON file:
```

File: firebase/app/routes/auth/google.tsx (hunk 8 above)
Comment (penx) [reply firebase/app/routes/auth/google.tsx:20]: confirmed this is an issue - using firebase functions locally (:5002) will redirect to :3000

File: firebase/app/routes/login.tsx [hunk 12]
```diff
@@ -113,6 +118,17 @@ export default function Login() {
           Login
         </button>
       </form>
+      <p>
+        <a
+          href={`https://accounts.google.com/o/oauth2/v2/auth\
+?response_type=code\
+&client_id=${GOOGLE_CLIENT_ID}\
+&redirect_uri=http://localhost:3000/auth/google\
```
Comment (penx) [inline firebase/app/routes/login.tsx:126]: also need to address hard coded localhost url here

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Comment by penx** [comment]
@MichaelDeBoey not sure why this was closed. Are you waiting for something from?

**Comment by penx** [comment]
Actually maybe it’s the hard coded localhost, had forgotten this needed addressing

**Comment by MichaelDeBoey** [comment]
@penx This was indeed waiting for the hardcoded `localhost` I think

I'll re-open as I know we'll see the update in a timely manner. 😉

I'll also update the `no-response` workflow to give a notice before actually closing the PRs 👍

**Comment by penx** [comment]
I have removed the 2 hard coded "localhost", and calculate redirectUri instead

File: firebase/README.md [hunk 13]
```diff
@@ -34,10 +37,23 @@ When you run `npm run emulators`, an initial user is created with credentials `u
 ## Deploying
 
 1. Follow the "Run against a Firebase Project" steps above if not done already
```
Comment (MichaelDeBoey) [inline firebase/README.md:39]:
```suggestion
1. Follow the "Run against a Firebase project" steps above if not done already
```

File: firebase/README.md [hunk 14]
```diff
@@ -34,10 +37,23 @@ When you run `npm run emulators`, an initial user is created with credentials `u
...
-3. Log in to the CLI with `firebase login`
-4. Run `firebase use --add` and choose the Firebase project you want to deploy to
-5. Deploy with `firebase deploy`
+2. Log in to the CLI with `npm run firebase -- login`
+3. Run `npm run firebase -- use --add` and choose the Firebase project you want to deploy to
+4. Deploy with `npm run firebase -- deploy`
+
+## Integration with Google Sign-in Provider
+
+The "Login with Google" link will only work when running against a Firebase Project.
```
Comment (MichaelDeBoey) [inline firebase/README.md:46]:
```suggestion
The "Login with Google" link will only work when running against a Firebase project.
```

File: firebase/README.md [hunk 15]
```diff
@@ -34,10 +37,23 @@ When you run `npm run emulators`, an initial user is created with credentials `u
...
+
+After the steps in "Run against a Firebase Project" have been completed:
```
Comment (MichaelDeBoey) [inline firebase/README.md:48]:
```suggestion
After the steps in "Run against a Firebase project" have been completed:
```

File: firebase/README.md [hunk 16]
```diff
@@ -34,10 +37,23 @@ When you run `npm run emulators`, an initial user is created with credentials `u
...
+
+- In the [Firebase Console](https://console.firebase.google.com), navigate to Authentication > Sign-in method > Add new provider > Google. Make a note of the client ID and secret and add them to the .env file as GOOGLE_CLIENT_ID and GOOGLE_CLIENT_SECRET.
```
Comment (MichaelDeBoey) [inline firebase/README.md:50]:
```suggestion
- In the [Firebase Console](https://console.firebase.google.com), navigate to Authentication > Sign-in method > Add new provider > Google. Make a note of the client ID and secret and add them to the `.env` file as `GOOGLE_CLIENT_ID` and `GOOGLE_CLIENT_SECRET`.
```

File: firebase/app/firebase-rest.ts [hunk 17]
```diff
@@ -54,9 +54,114 @@ export const signInWithPassword = async (
...
+export const signInWithIdp = async (
+  idToken: string,
+  providerId: string,
+  restConfig: {
+    apiKey: string;
+    domain: string;
+  }
+) => {
+  const body = {
+    postBody: "id_token=" + idToken + "&providerId=" + providerId,
```
Comment (MichaelDeBoey) [inline firebase/app/firebase-rest.ts:158]:
```suggestion
    postBody: `id_token=${idToken}&providerId=${providerId}`,
```

File: firebase/app/firebase-rest.ts [hunk 18]
```diff
@@ -54,9 +54,114 @@ export const signInWithPassword = async (
...
+    requestUri: "http://localhost",
```
Comment (MichaelDeBoey) [inline firebase/app/firebase-rest.ts:159]: @penx Should this stay `localhost`? 🤔

File: firebase/app/routes/login.tsx [hunk 19]
```diff
@@ -4,6 +4,7 @@ import {
   Link,
   useActionData,
   useLoaderData,
+  useLocation,
```
Comment (MichaelDeBoey) [inline firebase/app/routes/login.tsx:7]:
```suggestion
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]
Some small remarks

**Pull request closed without merging**